core/conn/spj_init/src/main/java/com/traf/mgmt/JarFileMgmt.java

```diff
@@ -0,0 +1,435 @@
+		sb.append("</rmRex>");
+		names[0] = sb.toString();
+		LOG.info("Done for removing files[" + pattern + "].");
+	}
+
+	public static void lsAll(String[] names) throws SQLException {
+		ls("*", names);
+	}
+
+	/**
+	 * list the Jars matching PATTERN
+	 * 
+	 * @param pattern:
+	 * @param names
+	 * @throws SQLException
+	 */
+	public static void ls(String pattern, String[] names) throws SQLException {
+		checkFileName(pattern);
+		Connection conn = getConn();
+		LOG.info("List files[" + pattern + "]");
+		String userPath = getCodeFilePath(conn);
+		close(conn);
+		File dir = new File(userPath);
+		if (!dir.exists() || !dir.isDirectory()) {
+			LOG.error("Directory [" + userPath + "] is not found!");
+			throw new SQLException("Directory [" + userPath + "] is not found!");
+		}
+		if (pattern == null) {
+			LOG.error("File pattern should not be empty!");
+			throw new SQLException("Pattern is empty!");
+		}
+		File[] files = getFiles(pattern, dir);
+		names[0] = toXML(files, "ls");
+	}
+
+	/**
+	 * upload a JAR file
+	 * 
+	 * @param fileData
+	 * @param fileName
+	 * @param appendFlag
+	 *            0: append; otherwise overwrite
+	 * @throws SQLException
+	 */
+	public static void put(String fileData, String fileName, int appendFlag) throws SQLException {
+		checkFileName(fileName);
+		try {
+			byte[] data = fileData.getBytes(CHARTSET);
+
+			Connection conn = getConn();
+			LOG.info("Put " + fileName + ", length: " + data.length + ", file string length:" + fileData.length());
+			String userPath = getCodeFilePath(conn);
+			close(conn);
+			String fname = userPath + fileName;
+			checkFile(fname, data.length);
+			FileOutputStream fos = null;
+			try {
+				fos = new FileOutputStream(fname, (appendFlag == 0));
+
+				fos.write(Snappy.uncompress(Arrays.copyOf(data, data.length)));
+				fos.flush();
+			} finally {
+				if (fos != null)
+					fos.close();
+			}
+
+			syncJar(userPath, fileName);
+
+		} catch (Throwable t) {
+			LOG.error(t.getMessage(), t);
+			throw new SQLException(t.getMessage());
+		}
+	}
+
+	private static void checkFileName(String fileName) throws SQLException {
+		if (fileName.contains("/") || fileName.contains("\\"))
+			throw new SQLException("Illegal file name: " + fileName
+					+ ". File name must not contain \"/\".");
+	}
+
+	private static void checkFile(String fname, int dataSize) throws SQLException {
+		File jar = new File(fname);
+		if (jar.length() + dataSize > MAX_JAR_FILE_SIZE) {
+			LOG.error("Jar file size is over the threshold[100Mb]");
+			throw new SQLException("Jar file size is over the threshold[100Mb]");
+		}
+	}
+
+	private static String getCodeFilePath(Connection conn) throws SQLException {
+		String user = getCurrentUser(conn);
+		String root = System.getenv("MY_SQROOT");
+		if (root == null || "".equals(root.trim())) {
+			LOG.error("Cant get your traf installation path!");
+			throw new SQLException("Cant get your traf installation path!");
+		}
+		File file = new File(root + "/spj/" + user);
```

Comment:
Can we make this root + "/udr/lib/" + user? This will be used for UDFs as well. Eventually, we want to replace this method with libraries that are stored in BLOBs and are cached, probably in directory root + "/udr/cache/".